module/applications/components/targeted-application-mixin.mjs

```diff
@@ -0,0 +1,144 @@
+     * @type {HTMLElement}
+     */
+    targetSourceControl;
+
+    /* -------------------------------------------- */
+    /*  Rendering                                   */
+    /* -------------------------------------------- */
+
+    /**
+     * Return the HTML elements needed to build the target source control and target list.
+     * @returns {HTMLElement[]}
+     */
+    buildTargetContainer() {
+      this.targetSourceControl = document.createElement("div");
+      this.targetSourceControl.classList.add("target-source-control");
+      this.targetSourceControl.innerHTML = `
+        <button type="button" class="unbutton" data-mode="targeted" aria-pressed="false">
+          <i class="fa-solid fa-bullseye" inert></i> ${game.i18n.localize("DND5E.Tokens.Targeted")}
+        </button>
+        <button type="button" class="unbutton" data-mode="selected" aria-pressed="false">
+          <i class="fa-solid fa-expand" inert></i> ${game.i18n.localize("DND5E.Tokens.Selected")}
+        </button>
+      `;
+      this.targetSourceControl.querySelectorAll("button").forEach(b =>
+        b.addEventListener("click", this._onChangeTargetMode.bind(this))
+      );
+      if ( !this.chatMessage?.getFlag("dnd5e", "targets")?.length ) this.targetSourceControl.hidden = true;
+
+      this.targetList = document.createElement("ul");
+      this.targetList.classList.add("targets", "unlist");
+
+      return [this.targetSourceControl, this.targetList];
+    }
+
+    /* -------------------------------------------- */
+
+    /**
+     * Build a list of targeted tokens based on current mode & replace any existing targets.
+     */
+    buildTargetsList() {
+      if ( !this.targetList ) throw new Error("Must create a element to contain the target list.");
+      let targetedTokens;
+      switch ( this.targetingMode ) {
+        case "targeted":
+          targetedTokens = (this.chatMessage?.getFlag("dnd5e", "targets") ?? [])
+            .map(t => ({ uuid: t.uuid, name: t.name }));
+          break;
+        case "selected":
+          targetedTokens = canvas.tokens?.controlled?.map(t => ({ uuid: t.actor.uuid, name: t.name })) ?? [];
+          break;
+      }
+      targetedTokens = Array.from(new Set(targetedTokens));
+      const targets = targetedTokens.map(t => this.buildTargetListEntry(t)).filter(t => t);
+      if ( targets.length ) this.targetList.replaceChildren(...targets);
+      else {
+        const li = document.createElement("li");
+        li.classList.add("none");
+        li.innerText = game.i18n.localize(`DND5E.Tokens.None${this.targetingMode.capitalize()}`);
+        this.targetList.replaceChildren(li);
+      }
+    }
+
+    /* -------------------------------------------- */
+
+    /**
+     * Create a list entry for a single target.
+     * @param {object} data
+     * @param {string} data.uuid  UUID of the targeted actor.
+     * @param {string} data.name  Name of the targeted token.
+     * @returns {HTMLLIElement|void}
```

Comment:
I think this must be implemented by any subclasses. If so, we should mark it `@abstract`.

```suggestion
     * @returns {HTMLLIElement|void}
     * @abstract
```